go/ephemeral/team_ek_box_storage.go

```diff
@@ -0,0 +1,374 @@
+	} else if found {
+		s.cache.PutMap(teamID, teamEKBoxes)
+	}
+	return teamEKBoxes, found, err
+}
+
+type TeamEKBoxedResponse struct {
+	Result struct {
+		Box              string                `json:"box"`
+		UserEKGeneration keybase1.EkGeneration `json:"user_ek_generation"`
+		Sig              string                `json:"sig"`
+	} `json:"result"`
+}
+
+func (s *TeamEKBoxStorage) fetchAndPut(ctx context.Context, teamID keybase1.TeamID, generation keybase1.EkGeneration) (teamEK keybase1.TeamEk, err error) {
+	defer s.G().CTrace(ctx, "TeamEKBoxStorage#fetchAndPut", func() error { return err })()
+
+	apiArg := libkb.APIArg{
+		Endpoint:    "team/team_ek_box",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args: libkb.HTTPArgs{
+			"team_id":    libkb.S{Val: string(teamID)},
+			"generation": libkb.U{Val: uint64(generation)},
+		},
+	}
+
+	var result TeamEKBoxedResponse
+	res, err := s.G().GetAPI().Get(apiArg)
+	if err != nil {
+		return teamEK, err
+	}
+
+	err = res.Body.UnmarshalAgain(&result)
+	if err != nil {
+		return teamEK, err
+	}
+
+	// Before we store anything, let's verify that the server returned
+	// signature is valid and the KID it has signed matches the boxed seed.
+	// Otherwise something's fishy..
+	teamEKStatement, wrongKID, err := verifySigWithLatestPTK(ctx, s.G(), teamID, result.Result.Sig)
+
+	// Check the wrongKID condition before checking the error, since an error
+	// is still returned in this case. TODO: Turn this warning into an error
+	// after EK support is sufficiently widespread.
+	if wrongKID {
+		s.G().Log.CWarningf(ctx, "It looks like you revoked a team key without generating new ephemeral keys. Are you running an old version?")
+		return teamEK, nil
+	}
+	if err != nil {
+		return teamEK, err
+	}
+
+	if teamEKStatement == nil { // shouldn't happen
+		s.G().Log.CWarningf(ctx, "No error but got nil teamEKMetadata")
+		return teamEK, err
+	}
+
+	teamEKMetadata := teamEKStatement.CurrentTeamEkMetadata
+	teamEKBoxed := keybase1.TeamEkBoxed{
+		Box:              result.Result.Box,
+		UserEkGeneration: result.Result.UserEKGeneration,
+		Metadata:         teamEKMetadata,
+	}
+
+	teamEK, err = s.unbox(ctx, teamEKBoxed)
+	if err != nil {
+		return teamEK, err
+	}
+
+	seed := TeamEKSeed(teamEK.Seed)
+	keypair := seed.DeriveDHKey()
+
+	if !keypair.GetKID().Equal(teamEKMetadata.Kid) {
+		return teamEK, fmt.Errorf("Failed to verify server given seed against signed KID %s", teamEKMetadata.Kid)
+	}
+
+	// Store the boxed version, return the unboxed
+	err = s.Put(ctx, teamID, generation, teamEKBoxed)
+	return teamEK, err
+}
+
+func (s *TeamEKBoxStorage) unbox(ctx context.Context, teamEKBoxed keybase1.TeamEkBoxed) (teamEK keybase1.TeamEk, err error) {
+	defer s.G().CTrace(ctx, "TeamEKBoxStorage#unbox", func() error { return err })()
+
+	userEKBoxStorage := s.G().GetUserEKBoxStorage()
+	userEK, err := userEKBoxStorage.Get(ctx, teamEKBoxed.UserEkGeneration)
+	if err != nil {
+		return teamEK, err
+	}
```

Comment:
This is going to be a common error path for people with stale devices, right? It's hard to know exactly what we'll need here until we start writing chat code, but maybe it's likely we'll want a specific error type, which the caller can check for?

Reply:
yeah i'll add a TODO if you get that error and it's > your max generation you need to make a new userEK